File: python-puka/rpc_server.py
#!/usr/bin/env python
import puka

client = puka.Client("amqp://localhost/")
promise = client.connect()
client.wait(promise)

promise = client.queue_declare(queue='rpc_queue')
client.wait(promise)

# The worlds worst algorithm:
def fib(n):
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fib(n-1) + fib(n-2)


print " [x] Awaiting RPC requests"
consume_promise = client.basic_consume(queue='rpc_queue', prefetch_count=1)
while True:
    msg_result = client.wait(consume_promise)
    n = int(msg_result['body'])

    print " [.] fib(%s)"  % (n,)
    response = fib(n)

    # This publish doesn't need to be synchronous.
    client.basic_publish(exchange='',
                         routing_key=msg_result['headers']['reply_to'],
                         headers={'correlation_id':
                                      msg_result['headers']['correlation_id']},
                         body=str(response))
    client.basic_ack(msg_result)

File: kotlin-gradle/src/main/kotlin/RPCServer.kt
import com.rabbitmq.client.AMQP
import com.rabbitmq.client.ConnectionFactory
import com.rabbitmq.client.DefaultConsumer
import com.rabbitmq.client.Envelope
import sun.jvm.hotspot.HelloWorld.fib


/**
 * Author: vincent
 * Date: 2018-06-08 17:35:00
 * Comment:
 */

class RPCServer {
    companion object {
        const val RPC_QUEUE_NAME = "rpc_queue"

        fun fib(n: Int): Int {
            if (n == 0) return 0
            return if (n == 1) 1 else fib(n - 1) + fib(n - 2)
        }
    }
}

fun main(args: Array<String>) {
    val factory = ConnectionFactory()
    factory.host = "localhost"
    val connection = factory.newConnection()
    val channel = connection.createChannel()

    channel.queueDeclare(RPCServer.RPC_QUEUE_NAME, false, false, false, null)

    channel.basicQos(1)

    System.out.println(" [x] Awaiting RPC requests")

    val consumer = object : DefaultConsumer(channel) {
        override fun handleDelivery(consumerTag: String, envelope: Envelope, properties: AMQP.BasicProperties, body: ByteArray) {
            val replyProps = AMQP.BasicProperties.Builder()
                    .correlationId(properties.correlationId)
                    .build()
            val message = String(body, charset("UTF-8"))
            val n = Integer.parseInt(message)
            println(" [.] fib($message)")
            val response = fib(n).toString()
            channel.basicPublish("", properties.replyTo, replyProps, response.toByteArray())
            channel.basicAck(envelope.deliveryTag, false)
        }
    }

    channel.basicConsume(RPCServer.RPC_QUEUE_NAME, false, consumer)
    // Wait and be prepared to consume the message from RPC client.
    while (true) {
        synchronized(consumer) {
            (consumer as Object).wait()
        }
    }
}